File: go/scrabble-score/scrabble_score.go
// Package scrabble Given a word, compute the Scrabble score for that word.
package scrabble

import (
	"regexp"
	"strings"
	"unicode"
)

// charScore returns an `int` score for a given letter (`string`).
func charScore(letter rune) int {
	letter = unicode.ToLower(letter)

	switch letter {
	case 'd', 'g':
		return 2
	case 'b', 'c', 'm', 'p':
		return 3
	case 'f', 'h', 'v', 'w', 'y':
		return 4
	case 'k':
		return 5
	case 'j', 'x':
		return 8
	case 'q', 'z':
		return 10
	default:
		// A, E, I, O, U, L, N, R, S, T
		return 1
	}
}

// Score returns an `int` scrablle score for a given word (`string`).
func Score(word string) int {
	var score int
	var letters string

	letters = strings.TrimSpace(word)

	// Make sure we have only have letters.
	if m, _ := regexp.MatchString(`^[a-zA-Z]+$`, word); !m {
		return 0
	}

	for _, letter := range letters {
		score += charScore(letter)
	}

	return score
}

File: dart/scrabble-score/lib/scrabble_score.dart
// Given a word, compute the Scrabble score for that word.
// https://exercism.org/tracks/dart/exercises/scrabble-score

// letterScore(String) returns the score of the letter
int letterScore(String letter) {
  letter = letter.toLowerCase();

  switch (letter) {
    case 'd':
    case 'g':
      return 2;
    case 'b':
    case 'c':
    case 'm':
    case 'p':
      return 3;
    case 'f':
    case 'h':
    case 'v':
    case 'w':
    case 'y':
      return 4;
    case 'k':
      return 5;
    case 'j':
    case 'x':
      return 8;
    case 'q':
    case 'z':
      return 10;
    default:
      // A, E, I, O, U, L, N, R, S, T
      return 1;
  }
}

// score(String) returns the score of the word
int score(String word) {
  word = word.toLowerCase();

  var result = 0;
  var letters = word.split('');

  for (final letter in letters) {
    result += letterScore(letter);
  }

  return result;
}
